src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <errno.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <sys/un.h>
+
+#define SD_LISTEN_FDS_START 3
+#define SOCKET_MODES 3
+#define MAX_COCKPIT_WS_ARGS 5
+
+int socket_to_pid[SOCKET_MODES];
+struct pollfd ufds[SOCKET_MODES];
+struct
+{
+  const char *sockname;
+  const char *argv[MAX_COCKPIT_WS_ARGS];
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static void
+handle_cockpit_ws_exited ()
+{
+  pid_t pid;
+  while ((pid = waitpid (-1, NULL, WNOHANG)) > 0)
+    {
+      for (int i = 0; i < SOCKET_MODES; i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = 0;
+            }
+        }
+    }
+}
+
+int
+spawn_cockpit_ws (char *ws_path, int index)
+{
+  int rv = 0;
+  char *args[MAX_COCKPIT_WS_ARGS + 1];
+  char *env[] = { NULL };
+
+  int i = 0;
+  for (i = 0; i < MAX_COCKPIT_WS_ARGS; i++)
+    {
+      if (instance_types[index].argv[i] != NULL)
+        args[i + 1] = strdup (instance_types[index].argv[i]);
```

Comment:
You don't need to `strdup()` here.  Declare the `args[]` variable as `const char *` and just assign the pointer directly.